tests/tests.hs

```diff
@@ -1,44 +1,145 @@
                                                generate)
 
-initializeDB :: MonadResource m => FilePath -> m DB
-initializeDB path =
-    open
-        path
-        defaultOptions
-        {createIfMissing = True, compression = NoCompression}
+testOpenOptions :: Options
+testOpenOptions =
+  defaultOptions
+    {createIfMissing = True, compression = NoCompression}
 
 main :: IO ()
 main =  hspec $ do
 
+  describe "Underlying library behaviour" $ do
+
+    it "runResourceTChecked doesn't ignore errors in finalizers" $ do
+      -- Added to `resourcet` so that our use case here is satisfied:
+      --   https://github.com/snoyberg/conduit/pull/347
+      runResourceTChecked $ do
+        _ <- allocate (return 'x') (\_x -> putStrLn "finalizer running" >> error "allocate finalizer fail")
+        return ()
+      `shouldThrow` \e -> case e of
+          ResourceCleanupException ex [] -> case fromException ex of
+            Just (ErrorCall "allocate finalizer fail") -> True
+            _ -> False
+          _ -> False
+
   describe "Basic DB Functionality" $ do
-    it "should put items into the database and retrieve them" $  do
-      runResourceT $ withSystemTempDirectory "rocksdb" $ \path -> do
-        db <- initializeDB path
+
+    it "should put items into the database and retrieve them (withDB way)" $ do
+      withSystemTempDirectory "rocksdb" $ \path -> do
+        withDB path testOpenOptions $ \db -> do
+          put db def "zzz" "zzz"
+          val <- get db def "zzz"
+          return val
+      `shouldReturn` (Just "zzz")
+
+    it "should put items into the database and retrieve them (non-async-safe way)" $ do
+      withSystemTempDirectory "rocksdb" $ \path -> do
+        db <- open path testOpenOptions
         put db def "zzz" "zzz"
-        get db def "zzz"
+        val <- get db def "zzz"
+        close db
+        return val
       `shouldReturn` (Just "zzz")
 
-    it "should put items into a database whose filepath has unicode characters and\
-       \ retrieve them" $  do
-      runResourceT $ withSystemTempDirectory "rocksdb" $ \path -> do
+    it "should put items into the database and retrieve them (MonadResource way)" $ do
+      runResourceTChecked $ withSystemTempDirectory "rocksdb" $ \path -> do
+        (_, db) <- openBracket path testOpenOptions
+        put db def "zzz" "zzz"
+        val <- get db def "zzz"
+        return val
+      `shouldReturn` (Just "zzz")
+
+    -- This test doesn't apply to Windows because there we can't delete the
+    -- directory while a file in it is open.
+#if !defined(mingw32_HOST_OS)
+    it "does weird global singleton string matching stuff for double-locking warnings" $ do
+      -- This test will fail when RocksDB fixes this issue
+      -- (see https://stackoverflow.com/questions/37310588/rocksdb-io-error-lock-no-locks-available#comment83145041_37312033).
+      -- It exists so that we get notified when that fixing happens.
```

Comment:
Is there anything we should do on our end when it happens? :)

Reply:
That depends on what they do instead. If they do what sqlite would probably do, then no. If they come up with some other way, you might want to inspect whether it's compatible with your way of using it.